jellyfin-core/src/main/kotlin/org/jellyfin/apiclient/discovery/RecommendedServerDiscovery.kt

```diff
@@ -0,0 +1,119 @@
+package org.jellyfin.apiclient.discovery
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.withContext
+import org.jellyfin.apiclient.Jellyfin
+import org.jellyfin.apiclient.api.operations.SystemApi
+import org.jellyfin.apiclient.model.api.PublicSystemInfo
+import org.jellyfin.apiclient.model.discovery.ServerVersion
+import org.slf4j.LoggerFactory
+import java.net.ConnectException
+
+class RecommendedServerDiscovery(
+	private val jellyfin: Jellyfin
+) {
+	private val logger = LoggerFactory.getLogger("RecommendedServerDiscovery")
+
+	private data class SystemInfoResult(
+		val address: String,
+		val systemInfo: PublicSystemInfo?,
+		val responseTime: Long
+	)
+
+	private suspend fun getSystemInfoResult(address: String): SystemInfoResult? {
+		logger.info("Requesting public system info for $address")
+
+		val client = jellyfin.createApi(baseUrl = address)
+		val api = SystemApi(client)
+
+		val startTime = System.currentTimeMillis()
+		val info = try {
+			api.getPublicSystemInfo()
+		} catch (err: ConnectException) {
+			logger.debug("Could not connect to $address")
+			null
+		} catch (err: Exception) {
+			logger.error("Could not retrieve public system info for $address", err)
+			null
+		}
+		val endTime = System.currentTimeMillis()
+
+		return SystemInfoResult(
+			address = address,
+			systemInfo = if (info != null && info.status == 200) info.content else null,
+			responseTime = endTime - startTime,
+		)
+	}
+
+	private fun assignScore(result: SystemInfoResult, parentResult: SystemInfoResult? = null): RecommendedServerInfo {
+		var points = 0
+
+		// Security
+		if (result.address.startsWith("https:")) points += 3
+
+		// Speed
+		when {
+			result.responseTime < 500 -> points += 3
+			result.responseTime < 1500 -> points += 2
+			result.responseTime < 5000 -> points += 1
+		}
+
+		// Compatibility
+		val version = result.systemInfo?.version?.let(ServerVersion::fromString)
+		if (version != null) {
+			points += 3
+
+			if (version < Jellyfin.apiVersion) points -= 2
+			if (version < Jellyfin.recommendedVersion) points -= 1
+		}
```

Comment:
Wouldn't this also work the other way around? Also, could you use a `when` instead?

Reply:
Can't use a `when` since both if-statements can be effective making the score `-3`. Inverting the logic might be possible.

Reply:
Well it depends, if `recommendedVersion` is always greater than `apiVersion`, you could simplify the statement and use a when. No need for two checks then as well.

Reply:
They might differ so I'll keep the if-statements. I did invert them though